alf/algorithms/vae.py

```diff
@@ -15,98 +15,149 @@
         self._beta = beta
 
-    def sampling_forward(self, inputs):
+    def _sampling_forward(self, inputs):
         """Encode the data into latent space then do sampling.
 
         Args:
-            inputs (Tensor or Tuple of Tensor): data to be encoded. If it has a prior network, then the argument
-            is a tuple of (prior_input, new_observation).
-        
+            inputs (nested Tensor): if a prior network is provided, this is a
+                tuple of `(prior_input, new_observation)`.
+
         Returns:
-            Tuple of z and kl_loss. z is tensor of shape (N, hidden_dim), kl_loss is tensor of shape (N,)
-        
+            (z, kl_loss): `z` is a tensor of shape (`N`, `z_dim`), `kl_loss` is
+                a tensor of shape (`N`,).
         """
-        if self._prior_network:
+        if self._z_mean_prior_network:
             prior_input, new_obs = inputs
-            prior_z_mean, prior_z_log_var = self._prior_network(prior_input)
-            inputs = tf.concat(
-                [prior_input, new_obs, prior_z_mean, prior_z_log_var], -1)
+            prior_z_mean, _ = self._z_mean_prior_network(prior_input)
+            prior_z_log_var, _ = self._z_log_var_prior_network(prior_input)
+            inputs = [prior_input, new_obs, prior_z_mean, prior_z_log_var]
 
-        latents = self._preprocess_layers(inputs)
+        latents, _ = self._preprocess_network(inputs)
         z_mean = self._z_mean(latents)
         z_log_var = self._z_log_var(latents)
 
-        if self._prior_network:
-            kl_div_loss = tf.square(z_mean) / tf.exp(prior_z_log_var) + \
-                          tf.exp(z_log_var) - z_log_var - 1.0
-
-            z_mean += prior_z_mean
-            z_log_var += prior_z_log_var
-
+        if self._z_mean_prior_network:
+            kl_div_loss = math_ops.square(z_mean) / torch.exp(prior_z_log_var) + \
+                          torch.exp(z_log_var) - z_log_var - 1.0
+            z_mean = z_mean + prior_z_mean
+            z_log_var = z_log_var + prior_z_log_var
         else:
-            kl_div_loss = tf.square(z_mean) + tf.exp(
+            kl_div_loss = math_ops.square(z_mean) + torch.exp(
                 z_log_var) - 1.0 - z_log_var
 
-        kl_div_loss = 0.5 * tf.reduce_sum(kl_div_loss, axis=-1)
+        kl_div_loss = 0.5 * torch.sum(kl_div_loss, dim=-1)
         # reparameterization sampling: z = u + var ** 0.5 * eps
-        eps = tf.random.normal(
-            tf.shape(z_mean), dtype=tf.float32, mean=0., stddev=1.0)
-        z = z_mean + tf.exp(z_log_var * 0.5) * eps
+        eps = torch.randn(z_mean.shape)
+        z = z_mean + torch.exp(z_log_var * 0.5) * eps
         return z, self._beta * kl_div_loss
+
+    def train_step(self, time_step: TimeStep, state):
+        """
+        Args:
+            time_step (TimeStep): input time step data, where its `observation`
+                is augmented with a prior input if there is a prior network.
```

Comment:
should make explicit how it is augmented: which one is first, which one is second